README.md
# Kirby3 SunCyclePages

**Requirement:** Kirby 3.0

## Coffee, Beer, etc.

This plugin is free but if you use it in a commercial project to show your support you are welcome to:
- [make a donation 🍻](https://www.paypal.me/omz13/10) or
- [buy me ☕☕☕](https://buymeacoff.ee/omz13) or
- [buy a Kirby license using this affiliate link](https://a.paddle.com/v2/click/1129/36191?link=1170)

### Why call it suncycle?

There are perhaps, with the benefit of hindsight, better names such as 'embargo', 'autopublish', 'schedule', etc. However, when this plugin was developed the notion was to follow the lifecycle of a page: sunrise and sunset are well-known terms within SDLC or ILM (although very much less familiar outside this domain), and 'suncycle' was used. I had my head in software mode not publishing or journalism mode; it could also be argued that unless you are familiar with publishing and journalism, terms such as 'embargo' and 'schedule' are also unusual outside that domain.

## Documentation

### Purpose

For a kirby3 site, this plugin [omz13/suncyclepages](https://github.com/omz13/kirby3-suncyclepages) allows more granular control over the _lifecycle_ of a page to be achieved vis-à-vis pages becoming _generally available_ at a "sunrise date" and _withdrawn_ at a "sunset date". This enhances the limited lifecycle options provided by kirby3's `state` (unpublished, draft, and published) to better match that needed certain editorial or regulatory needs, _viz._, unpublished, draft, embargoed (waiting "sunrise"), published ("sunny"), and finally withdrawn ("sunset").

When would you use this plugin?

- The main use case is for writing blog posts that you want to publish in the future. Write them now, and schedule them to appear (sunrise) in the future. Ideal for those times when you are away but want to continue a publishing schedule.
- Another use case is for temporary offers or notices. You publish a page, but in the future you want it removed (sunset), or better yet, want it removed and replaced (redirected) to elsewhere that says "too late" (or whatever).

The functional specification:

- Only pages that have a status of "published" are affected, i.e. those with "draft" or "unpublished" behave as usual.
- Pages can be embargoed until being made generally available a specified date ("sunrise"). Any attempt to view the page before that date will yield the standard error page and a `404` [Not Found] code.
- Pages can be withdrawn at a specific date ("sunset"). Any attempt to view the page after that date will yield a `410` [Gone] and the standard error page, or, a `301` [Moved Permanently] to a specified location on a per-page basis.
- A debug mode can be enabled to include a header in the response to indicate that a page was under embargo or has sunset, c.f. `X-SUNCYCLE`.

#### Roadmap

The non-binding list of planned features and implementation notes are:

- [x] MVP
- [x] Basic debugging
- [ ] If the time is not specified (just the date), default sensibly.
- [ ] Better debugging
- [x] Blueprints

### Installation

#### via composer

If your kirby3-based site is managed using-composer, simply invoke `composer require omz13/kirby3-suncyclepages`, or add `omz13/kirby3-suncyclepages` to the "require" component of your site's `composer.json` as necessary, e.g. to be on the bleeding-edge:

```yaml
"require": {
  ...
  "omz13/kirby3-suncyclepages": "dev-master as 1.0.0",
  ...
}
```
#### via git

Clone github.com/omz13/kirby3-suncyclepages into your `site/plugins` and then in `site/plugins/kirby3-suncyclepages` invoke ``composer update --no-dev`` to generate the `vendor` folder and the magic within.

```sh
$ git clone github.com/omz13/kirby3-suncyclepages site/plugins/kirby3-suncyclepages
$ cd site/plugins/kirby3-suncyclepages
$ composer update --no-dev
```

If your project itself is under git, then you need to add the plugin as a submodule and possibly automate the composer update; it is assumed if you are doing this that you know what to do.

### Configuration

The following mechanisms can be used to modify the plugin's behavior.

#### via `config.php`

- `omz13.suncyclepages.disable` - optional - default `false` - a boolean which, if `true`, disables the plugin.

- `omz13.suncyclepages.embargoCheckWhenTemplateIs` - optional - default `[ 'article' ]` : if a page uses a template name that is specified in this option, the embargo check (sunrise) is explicitly performed.

- `omz13.suncyclepages.embargoCheckWhenParentIs` - optional - default `[ 'blog' ]` : if a page has a parent that is specified in this parameter, the embargo check (sunrise) is explicitly performed.

### Use

#### Content fields

The plugin uses the following content fields. These are all optional; if missing or empty, they are assumed to be not applicable vis-à-via their indicated functionality.

- `date` - date - optional - the date when the page is to be published, i.e. embargo until this date. If the date field is missing or cannot be resolved, the embargo check will not embargo the page, i.e. failsafe is to not embargo but publish.

- `skipembargo` - boolean - optional (default `false`) - if `true`, the embargo check (sunrise) is skipped (not performed) for this page.

- `embargo` - boolean -  optional (default `false`) - gf `true`, the embargo check (sunrise) should be explicitly performed against this page. This is intended to be used on pages where such checking is not normally done, i.e. not explicitly done due to parent (c.f. `omz13.suncyclepages.embargoCheckWhenParentIs`) or template (c.f. `omz13.suncyclepages.embargoCheckWhenTemplateIs`).

- `sunset` - date - optional - the (future) date after which the page is to be sunset (withdrawn).

- `sunsetto` page - optional - the name of a page to be used for `301` redirects when the page is sunset; if not specified a `404` is given.

#### Blueprints

This plugin provides the following built-in blueprints (e.g. to make adding fields into the panel's blueprint `blueprint/site.yml` easier):

- `omz13/suncycle` - a group of fields to be used as an `extends` to make entering embrgo parameters easy. For example, it would be appened into starterkit's `site/blueprints/pages/article.yml` so:

```yaml
sidebar:
  meta:
    type: fields
    fields:
      date:
        type: date
        time: true
        default: now
        label: Publication Date
      author:
        type: users
      tags:
        type: tags
        options: query
        query:
          fetch: site.tags.toStructure.sortBy("name", "asc")
          text: "{{ structureItem.name }}"
          value: "{{ structureItem.value }}"

      extends: omz13/suncycle
```

This built-in blueprint is localized for `en` and `fr`; PRs for others are welcome!

### Use

#### Debug mode

If the kirby site is in debug mode:

- If a page is embargoed (waiting for sunrise), the `404` page will include an additional header `X-SUNCYCLE: isUnderembargo`.

- If a page has been withdrawn (sunset), the `410` response will include an additional header `X-SUNCYCLE: isSunset`, and the `301` a `X-SUNCYCLE: isSunset <to>` where `<to>` is the name of the `sunsetto` page.

#### Example Use in a collection

In the kirby3 StarterKit you don't want the list of blog post to include any that are under embargo (waiting sunrise) or are sunset... it is very simple to implement.

`site/collections/articles.php` looks like this:

```php
<?php

return function ($site) {
    return $site->find('blog')->children()->listed()->flip();
};
```

Change it to:
```php
<?php

return function ($site) {
	    return $site->find('blog')->children()->listed()->isunderembargo(false)->issunset(false)->flip();
};
```

Note the use of `false` in the filters for `isunderembargo` and `issunset` because its the pages that are NOT in this condition that are required.

#### Methods

This plugin makes the following methods available:

- `page.issunset()` : returns a boolean which is `true` if the page is currently sunset (i.e. withdrawn).
- `page.isunderembargo()` : returns a boolean which is `true` if the page is currently under embargo (i.e. waiting for sunrise).

- `pages.issunset( $match = true )` : a filter to return the subset of pages in a collection that are ( `$match = true` ) or are not ( `$match = false` ) sunset (i.e. withdrawn).
- `pages.isunderembargo( $match = true )` : a filter to return the subset of pages in a collection that are ( `$match = true` ) or are not ( `$match = false` ) under embargo (i.e. waiting for sunrise).

## Disclaimer

This plugin is provided "as is" with no guarantee. Use it at your own risk and always test it yourself before using it in a production environment. If you find any issues, please [create a new issue](https://github.com/omz13/kirby3-suncyclepages/issues/new).

## License

[MIT](https://opensource.org/licenses/MIT)

You are prohibited from using this plugin in any project that promotes racism, sexism, homophobia, animal abuse, violence or any other form of hate speech.


CHANGELOG

VERSION 1  RELEASE
==================

   Version 1.1 - Release
      15/02/2019 20:57  1.1.0  initial release
         bc4a330 Feat: install by zip or submodule

   Version 1.0 - Release
      15/01/2019 20:00  1.0.0  initial release
         f7fe409 Chore: Kirby 3.0

VERSION 0  ALPHA RELEASE
========================

   Version 0.4 - beta
      09/01/2019 02:10  0.4.0  initial release
         c1e7074 Docs: Coffee, Beer, etc
         109002e Chore: stan
         84a3ecc Feat: helper blueprint
         d822e10 Chore: toolchain fix
         881d547 Chore: Kirby RC2 or better, composer update
         c7d1f94 Docs: edit stupid mistakes
         ed812f2 Fixed micro typo

   Version 0.3 - beta release
      08/11/2018 17:25  0.3.0  initial release
         b70321f Chore: composer update
         fc1d282 Fix: update guard to cope with ML site
         95f5983 Docs: Add use cases. Closes #1.
         0adc299 Docs: edits for clarity

   Version 0.2 - beta release
      26/09/2018 18:40  0.2.1  beta release
         e3587c5 Chore: refactor composer
         344c07b Docs: detail debug headers
         b52f138 Fix: guard
      23/09/2018 14:44  0.2.0  initial release
         ec1a32a Chore: update lock
         e3656f8 Feature: embargoCheckWhenTemplateIs, embargoCheckWhenParentIs
         e4d5c81 Fix: add guard
         173f5a2 Chore: stan goes strict and tweak release process
         1d31cc7 Chore: squash stan false positives
         1708040 Fix: show raw to in header
         a292e61 Chore: guard
         f32678c Chore: style
         930fa24 Fix: no need for 418 (see header), so always 404
         921b975 Chore: ignore $this related errors until phpstan understands magic
         6f94c08 Chore: because I can’t type composer, use make
         a360187 Refactor: phpstan happiness and style update
         889a0af Chore: automagically normalize composer.json
         b4c7656 Chore: update cs
         78440ea Chore: update and apply coding ruleset

   Version 0.1 - alpha release
      31/08/2018 17:21  0.1.0  initial release

src/suncyclepages.php
<?php

// phpcs:disable Squiz.Commenting.ClassComment.Missing
// phpcs:disable Squiz.Commenting.VariableComment.Missing
// phpcs:disable Squiz.Commenting.FunctionComment.Missing

namespace omz13;

use const SUNCYCLE_CONFIGURATION_PREFIX;
use const SUNCYCLE_VERSION;

use function array_key_exists;
use function define;
use function is_array;
use function kirby;

define( 'SUNCYCLE_VERSION', '1.1.0' );
define( 'SUNCYCLE_CONFIGURATION_PREFIX', 'omz13.suncyclepages' );


class SunCyclePages
{

  public static function version() : string {
    return SUNCYCLE_VERSION;
  }//end version()

  public static function ping() : string {
      return static::class . ' pong ' . static::version();
  }//end ping()

  public static function getConfigurationForKey( string $key, ?string $default = null ) : ?string {
    $o = kirby()->option( SUNCYCLE_CONFIGURATION_PREFIX . '.' . $key );

    if ( isset( $o ) == true ) {
      return $o;
    }

    return $default;
  }//end getConfigurationForKey()

  public static function getArrayConfigurationForKey( string $key ) : array {
    // Try to pick up configuration when provided in an array (vendor.plugin.array(key=>value))
    $o = kirby()->option( SUNCYCLE_CONFIGURATION_PREFIX );
    if ( $o != null && is_array( $o ) && array_key_exists( $key, $o ) ) {
      return $o[$key];
    }

    // try to pick up configuration as a discrete (vendor.plugin.key=>value)
    $o = kirby()->option( SUNCYCLE_CONFIGURATION_PREFIX . '.' . $key );
    if ( $o != null ) {
      return $o;
    }

    // this should not be reached... because plugin should define defaults for all its options...
    return [];
  }//end getArrayConfigurationForKey()

  public static function isEnabled() : bool {
    if ( static::getConfigurationForKey( 'disable' ) == true ) {
      return false;
    } else {
      return true;
    }
  }//end isEnabled()
}//end class


src/config.php
<?php

// phpcs:disable PHPCompatibility.PHP.NewClosure.ThisFoundOutsideClass

Kirby::plugin(
    'omz13/suncyclepages',
    [
      'root'         => dirname( __FILE__, 2 ),
      'options'      => [
        'disable' => false,
        'embargoCheckWhenTemplateIs' => [ 'article', 'pfo-article' ],
        'embargoCheckWhenParentIs' => [ 'blog' ],
      ],
      'blueprints' => [
        'omz13/suncycle' => dirname( __FILE__, 2 ) . '/blueprints/suncycle.yml',
      ],

      'hooks'        => [
        // phpcs:ignore
        'route:after' => function ( $route, string $path, string $method, $result ) {

          if ( $route->env() != 'site' ) {
              return;
          }

          if ( $result == null ) { // hmmm... requested page not found (not in kirby content)
            return;
          }

          if ( ! ( $result instanceof Kirby\Cms\Page ) ) {
            // Ignore it because not a page; very probably a Kirby\Cms\Response
            return;
          }

          if ( omz13\SunCyclePages::isEnabled() == false ) {
              return false;
          }

          if ( property_exists( $result, 'content' ) == false ) {
              return;
          }

          // belt-and-braces guarding
          if ( $result->hasMethod( 'issunset' ) == true ) {
            if ( $result->callMethod( 'issunset' ) == true ) {
              if ( $result->content()->has( 'sunsetto' ) ) {
                $to = $result->content()->get( 'sunsetto' );
                if ( kirby()->option( 'debug' ) == 'true' ) {
                  header( 'X-SUNCYCLE: isSunset to ' . $to );
                }
                if ( $to != "" ) {
                  go( $to, 301 );
                }
              }

              // because
              header( 'X-SUNCYCLE: isSunset' );
              // 410 = Gone.
              echo Kirby\Cms\Response::errorPage( [], 'html', 410 );
              die;
            }//end if
          }//end if

          if ( $result->hasMethod( 'isunderembargo' ) == true ) {
            if ( $result->callMethod( 'isunderembargo' ) == true ) {
              if ( kirby()->option( 'debug' ) == 'true' ) {
                header( 'X-SUNCYCLE: isUnderembargo' );
              }

              echo Kirby\Cms\Response::errorPage( [], 'html', 404 );
              die;
            }
          }
        },
      ],
      'pageMethods'  => [
        'issunset'       => function () {
          if ( omz13\SunCyclePages::isEnabled() == false ) {
            return false;
          }

          $timestamp = strtotime( $this->content()->get( 'sunset' ) );
          if ( $timestamp != 0 && $timestamp < time() ) {
            return true;
          }

          return false;
        },
        'isunderembargo' => function () {
          if ( omz13\SunCyclePages::isEnabled() == false ) {
            return false;
          }

          assert( $this instanceof Kirby\Cms\Page );

          if ( $this->content()->get( 'skipembargo' ) == 'true' ) {
            return false;
          }
          if ( in_array( $this->parent(), omz13\SunCyclePages::getArrayConfigurationForKey( 'embargoCheckWhenParentIs' ), false ) == true
            || in_array( $this->intendedTemplate(), omz13\SunCyclePages::getArrayConfigurationForKey( 'embargoCheckWhenTemplateIs' ), false ) == true
            || $this->content()->get( 'embargo' ) == 'true'
            ) {
            $timestamp = strtotime( $this->content()->get( 'date' ) );
            if ( $timestamp != 0 && time() < $timestamp ) {
              return true;
            }
          }

            return false;
        },
      ],
      'pagesMethods' => [
        'isunderembargo' => function ( $match = true ) {
          if ($match) { // phpcs:ignore
            return $this->filterBy( 'isunderembargo', true );
          }

          return $this->filterBy( 'isunderembargo', '!=', true );
        },
        'issunset'       => function ( $match = true ) {
          if ($match) {  // phpcs:ignore
            return $this->filterBy( 'issunset', true );
          }

          return $this->filterBy( 'issunset', '!=', true );
        },
      ],
      'collections'  => [
        'isunderembargo' => function ( $site ) {
          return $site->index()->isunderembargo();
        },
        'issunsetted'    => function ( $site ) {
          return $site->index()->issunset();
        },
      ],
    ]
);

require_once __DIR__ . '/suncyclepages.php';
